lightningd/peer_control.c

```diff
@@ -248,11 +253,83 @@ static bool get_peer_gossipfd_reply(struct subd *subd, const u8 *msg,
 	return true;
 }
 
-static void get_gossip_fd_for_reconnect(struct lightningd *ld,
-					const struct pubkey *id,
-					u64 unique_id,
-					int peer_fd,
-					const struct crypto_state *cs)
+static void get_gossip_fd_for_channeld_reconnect(struct lightningd *ld,
+						 const struct pubkey *id,
+						 u64 unique_id,
+						 int peer_fd,
+						 const struct crypto_state *cs)
+{
+	struct getting_gossip_fd *ggf = tal(ld, struct getting_gossip_fd);
+	u8 *req;
+
+	ggf->peer_fd = peer_fd;
+	ggf->id = *id;
+	ggf->cs = *cs;
+
+	/* FIXME: set sync to `initial_routing_sync` */
+	req = towire_gossipctl_get_peer_gossipfd(ggf, unique_id, true);
+	subd_req(ggf, ld->gossip, take(req), -1, 1,
+		 get_peer_gossipfd_channeld_reply, ggf);
+}
+
+static bool get_peer_gossipfd_closingd_reply(struct subd *subd, const u8 *msg,
```

Comment:
Yes, I disliked this immensely.  But I think the real fix may be:

1. gossipd listens to the socket for new connections, and also handles outgoing, absorbing
    handshaked.
2. gossipd thus knows who we're trying to connect to, and knows to return them immediately.

That unifies the paths nicely and removes all this code.  When gossipd hands a connection
to us, it gives us the gossip fd for that peer, and it's all good.

I've put in a FIXME to that effect.


 which gets told what peers we're trying to connect to.